libr/bin/format/mach0/coresymbolication.c

```diff
@@ -0,0 +1,306 @@
+#include <r_types.h>
+#include <r_util.h>
+#include <r_hash.h>
+#include "coresymbolication.h"
+
+#define R_CS_EL_OFF_SEGS 0x58
+#define R_CS_EL_SIZE_SEG 0x20
+#define R_CS_EL_SIZE_SECT_64 0x18
+#define R_CS_EL_SIZE_SECT_32 0x10
+#define R_CS_EL_SIZE_SYM 0x18
+#define R_CS_EL_SIZE_LSYM 0x24
+#define R_CS_EL_SIZE_LINFO 0x14
+
+static RCoreSymCacheElementHdr *r_coresym_cache_element_header_new(RBuffer *buf, size_t off, int bits) {
+	RCoreSymCacheElementHdr *hdr = R_NEW0 (RCoreSymCacheElementHdr);
+	if (!hdr) {
+		return NULL;
+	}
+	if (r_buf_fread_at (buf, off, (ut8 *)hdr, "13i16c5i", 1) != sizeof (RCoreSymCacheElementHdr)) {
+		free (hdr);
+		return NULL;
+	}
+	return hdr;
+}
+
+void r_coresym_cache_element_free(RCoreSymCacheElement *element) {
+	if (!element) {
+		return;
+	}
+	free (element->hdr);
+	free (element->file_name);
+	free (element->binary_version);
+	free (element->segments);
+	free (element->sections);
+	free (element->symbols);
+	free (element->lined_symbols);
+	free (element->line_info);
+	free (element);
+}
+
+ut64 r_coresym_cache_element_pa2va(RCoreSymCacheElement *element, ut64 pa) {
+	size_t i;
+	for (i = 0; i < element->hdr->n_segments; i++) {
+		RCoreSymCacheElementSegment *seg = &element->segments[i];
+		if (seg->size == 0) {
+			continue;
+		}
+		if (seg->paddr < pa && pa < seg->paddr + seg->size) {
+			return pa - seg->paddr + seg->vaddr;
+		}
+	}
+	return pa;
+}
+
+static void meta_add_fileline(RBinFile *bf, ut64 vaddr, ut32 size, RCoreSymCacheElementFLC *flc) {
+	Sdb *s = bf->sdb_addrinfo;
+	if (!s) {
+		return;
+	}
+	char aoffset[64];
+	ut64 cursor = vaddr;
+	ut64 end = cursor + R_MAX (size, 1);
+	char *fileline = r_str_newf ("%s:%d", flc->file, flc->line);
+	while (cursor < end) {
+		char *aoffsetptr = sdb_itoa (cursor, aoffset, 16);
+		if (!aoffsetptr) {
+			return;
+		}
+		if (!sdb_add (s, aoffsetptr, fileline, 0)) {
```

Comment:
sdb_add adds if not set, and set adds even if its there, this seems to me like dupe